src/cpp/transport/tcp/TCPServerLinux.cpp

```diff
@@ -150,53 +150,70 @@ bool TCPServer::send_message(OutputPacket output_packet)
         TCPConnection& connection = connections_.at(it->second);
         lock.unlock();
 
-        /* Send message size. */
         msg_size_buf[0] = uint8_t(0x00FF & output_packet.message->get_len());
         msg_size_buf[1] = uint8_t((0xFF00 & output_packet.message->get_len()) >> 8);
+        uint8_t n_attemps = 0;
+        uint16_t bytes_sent = 0;
+
+        /* Send message size. */
+        bool size_sent = false;
         do
         {
             uint8_t errcode;
-            send_rv = send_locking(connection, msg_size_buf, 2, errcode);
+            size_t send_rv = send_locking(connection, msg_size_buf, 2, errcode);
             if (0 < send_rv)
             {
                 bytes_sent += uint16_t(send_rv);
+                size_sent = (bytes_sent == 2);
             }
             else
             {
                 if (0 < errcode)
                 {
-                    close_connection(connection);
-                    rv = false;
+                    break;
                 }
             }
+            ++n_attemps;
         }
-        while (rv && bytes_sent != 2);
+        while (!size_sent && n_attemps < max_attemps);
 
         /* Send message payload. */
-        if (rv)
+        bool payload_sent = false;
+        if (size_sent)
         {
+            n_attemps = 0;
             bytes_sent = 0;
             do
             {
                 uint8_t errcode;
-                send_rv = send_locking(connection,
-                                       output_packet.message->get_buf() + bytes_sent,
-                                       output_packet.message->get_len() - bytes_sent,
-                                       errcode);
+                size_t send_rv = send_locking(connection,
+                                              output_packet.message->get_buf() + bytes_sent,
+                                              output_packet.message->get_len() - bytes_sent,
+                                              errcode);
                 if (0 < send_rv)
                 {
                     bytes_sent += uint16_t(send_rv);
+                    payload_sent = (bytes_sent == uint16_t(output_packet.message->get_len()));
                 }
                 else
                 {
                     if (0 < errcode)
                     {
-                        close_connection(connection);
-                        rv = false;
+                        break;
                     }
                 }
+                ++n_attemps;
             }
-            while (rv && bytes_sent != uint16_t(output_packet.message->get_len()));
+            while (!payload_sent && n_attemps < max_attemps);
+        }
+
+        if (size_sent && payload_sent)
```

Comment:
if `payload_sent` is true, is because `size_sent` is already true, so the line can be simplified to
```c
if (payload_sent)
```

Reply:
You are right! Done.